lib/Core/TimingSolver.h

```diff
@@ -10,60 +10,66 @@
 #ifndef KLEE_TIMINGSOLVER_H
 #define KLEE_TIMINGSOLVER_H
 
+#include "klee/Constraints.h"
 #include "klee/Expr.h"
-#include "klee/Solver.h"
 #include "klee/Internal/System/Time.h"
+#include "klee/Solver.h"
 
 #include <memory>
 #include <vector>
 
 namespace klee {
-  class ExecutionState;
```

Comment:
I don't like this dropping of `ExecutionState` from the `TimingSolver`. While I agree that having ExecutionState in the TimingSolver is annoying, this approach is not better. Essentially what this does is unpacks state the TimingSolver needs into constraints and this metadata, which only solves the problem of TimingSolver seeing too much execution state. However currently I see this as a bit of a feature, because there are cases where you need execution state for query solving and this is an easy point to introduce it, these changes make it harder. In addition the new TimingSolver still doesn't conform to the Solver API.

Therefore I would suggest to drop the metadata parameter, make TimingSolver conform to the Solver API and use some form of depedency inversion to get the queryCost metadata. I can see two ways of doing this, either make the queryCost update somewhere else, like `Executor::processTimers`, where you read the query cost for this state add it to state and reset it to zero. Alternatively you could have timing solvers for each state and the TimingSolver would then update the state similar to how it does it now. I guess the first proposal might be better.

Reply:
The goal is to separate executor and solver chain. This way, the timing solver can be reused in `kleaver` eventually. Second, if you need, you can add a pointer to the state as part of the meta data (which can be a nullptr) but you don't need to construct one.

I agree the TimingSolver should be made conform but I didn't want to overload this PR even more. The next step is to provide the metadata also to `Solver(Impl)` and streamline the solvers.

`Executor::processTimers` is not the right place, because you cannot correctly assign meta data to the correct state, i.e. what if you handle multiple states and their solver calls between `Executor::processTimers`. The other prosed alternative would involve creating timing solvers for each states which I would like to avoid in the first place.